### DISCUSSION TITLE
DeployLottery.s.sol -->  failed: panic: arithmetic underflow or overflow

### DISCUSSION THREAD

**Discussion by saikrishna5907** [Cyfrin/foundry-full-course-cu#1956] [Q&A]
Hi,

Anyone facing this issue when we try to deploy the raffle contract on anvil 

```
forge script script/DeployLottery.s.sol --rpc-url http://localhost:8545 --account local_anvil --broadcast
```

error screenshot
![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/39937402/c8e06227-bf59-498a-9005-6b4d4c5c7f08)

when I did some digging found out VRFCoordinatorV2_5Mock is extending SubscriptionAPI (from lib/chainlink-brownie-contracts/contracts/src/v0.8/vrf/dev/SubscriptionAPI.sol)

the error is causing by line number 355
![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/39937402/bd28c82b-2db9-46da-82d9-10f3449e4509)

particularly 
```
blockhash(block.number - 1)
```
here the block.number is 0, so when it evaluates (0 - 1) it is giving the error ```script failed: panic: arithmetic underflow or overflow (0x11)```

I did tried to change/ fix in SubscriptionAPI by adding this line 
```
bytes32 blockHashToUse = block.number > 0 ? blockhash(block.number - 1) : bytes32(0);
```
![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/39937402/a033cab6-3c5d-4a9c-a8b8-df3ec602a8be)

but now I get a new error on addConsumer function 
![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/39937402/a4c4b961-5fdb-449e-bb24-4ca54ecd9df9)

coming from here 
![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/39937402/10403abb-b236-4eff-bbce-cc0dd23eccaa)

Can anyone help please?

**Comment by EngrPips** [comment]
Hello @saikrishna5907, Please show your createSubscription script.

**Reply by saikrishna5907** [reply]
Hi @EngrPips here is my repo https://github.com/saikrishna5907/lottery-contract

here is my deploy script
![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/39937402/d5f7e3da-8180-423d-ab69-eebf8674dab4)

CreateSubscription contract
![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/39937402/3421cd75-cf59-4048-ba37-c50234fe6881)

**Reply by EngrPips** [reply]
Thanks for sharing the repo, I will take a look and point out whatever i find that i think can be causing the issue.

**Comment by EngrPips** [comment]
Hello @saikrishna5907, I want us to know that whenever you use `vm.startBroadcast()` in foundry and you did not pass in a `private-key`, Foundry will automatically use a default private key. Now that this fact is establish lets look into your code and see what is wrong

Here in your `createSubscription` script you have the below function.

```js
function createSubscription(
        address vrfCoordinator,
        address account
    ) public returns (uint256, address) {
        // create a subscription
        console2.log(
            "Creating a subscription on ChainId:",
            block.chainid,
            account
        );

        vm.startBroadcast();
        uint256 subscriptionId = VRFCoordinatorV2_5Mock(vrfCoordinator)
            .createSubscription();
        vm.stopBroadcast();
        console2.log("Your Subscription Id:", subscriptionId);
        return (subscriptionId, vrfCoordinator);
    }
```
This function uses `vm.startBroadcast()` and doesn't pass in any private key, which makes it so that it is the `Default Foundry Account` that creates the subscription. 

So when it got to the point where you wanted to add a consumer to that subscription you use the function below.

```js
function addConsumer(
        address mostRecentlyDeployedLottery,
        address vrfCoordinator,
        uint256 subscriptionId,
        address account
    ) public {
        console2.log(
            "Adding contract:",
            mostRecentlyDeployedLottery,
            "as a consumer"
        );
        console2.log("Using vrfCoordinator:", vrfCoordinator);
        console2.log("Using Subscription:", subscriptionId);
        console2.log("On ChainId:", block.chainid);
        vm.startBroadcast(account);
        VRFCoordinatorV2_5Mock(vrfCoordinator).addConsumer(
            subscriptionId,
            mostRecentlyDeployedLottery
        );
        vm.stopBroadcast();
    }
```
The `addConsumer` clearly uses `vm.startBroadcast(account)` passing in the private key of some account named `account`, doing that makes it so it the `account` and not the `Default Foundry Account` that is sending the account. The subId doesn't recognize this `account` private key so it reverts with `mustBeSubOwner`

**Suggested Solution**
Have your `createSubscription` as the below and re-run the test, Let see if it works

```js
function createSubscription(
        address vrfCoordinator,
        address account
    ) public returns (uint256, address) {
        // create a subscription
        console2.log(
            "Creating a subscription on ChainId:",
            block.chainid,
            account
        );

        vm.startBroadcast(account);
        uint256 subscriptionId = VRFCoordinatorV2_5Mock(vrfCoordinator)
            .createSubscription();
        vm.stopBroadcast();
        console2.log("Your Subscription Id:", subscriptionId);
        return (subscriptionId, vrfCoordinator);
    }
```

**Reply by saikrishna5907** [reply]
Thank you very much for spotting that @EngrPips 

Oh yeah... totally missed to pass the account param to startBraodcast() there.

I have just tried adding that and ran the deploy script it is deployed now, but got this error
![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/39937402/9027d93c-4c1a-4a4b-a27d-df03e70a9f8b)


I guess that is okay i think as I am using the default foundry address used in Base.sol
![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/39937402/6c68f2c0-d660-4454-90bf-f0201342be64)

**Reply by EngrPips** [reply]
Interesting, This is the first time I will be seeing this error lol. But everything works fine, right?

**Reply by saikrishna5907** [reply]
Hi @EngrPips my bad, when I said previously it was working its because I has this line at lib/chainlink-brownie-contracts/contracts/src/v0.8/vrf/dev/SubscriptionAPI.sol

```
bytes32 blockHashToUse = block.number > 0 ? blockhash(block.number - 1) : bytes32(0);
```
if I remove that I get same error again 
![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/39937402/76ee6648-20f3-4fca-8b2f-3ca5ca6c18d6)

**Reply by EngrPips** [reply]
I didn't need to add any line to any library code for my deployment script to work, So it is strange to me that you needed to add a line, but if that is what it takes for it to work for you, keep the line there till you figure how to resolve it. Re-run your test and send the test output. Please copy the output and paste it here instead of sending pictures. Don't forget to format it properly. Also, comment out your magic line, so the deployment will fail.

**Reply by saikrishna5907** [reply]
tests was never an issue deploying using the script is what causing the error 

here is the test output after commenting the fix added to lib
```
Ran 21 tests for test/unit/Lottery.t.sol:LotteryTest
[PASS] testCannotEnterLotteryWhenWinnerIsCalculating() (gas: 221724)
[PASS] testCheckUpKeepIfEnoughNotPassed() (gas: 18963)
[PASS] testCheckUpKeepRetunrsTrueWhenAllConditionsAreMet() (gas: 75121)
[PASS] testCheckUpKeepReturnsFalseIfNoBalance() (gas: 22043)
[PASS] testCheckUpKeepReturnsFalseIfThereAreNoPlayersYet() (gas: 20108)
[PASS] testCheckUpKeepReturnsFalseWhenNotOpen() (gas: 216537)
[PASS] testEmitsEnteredLotteryEventWhenPlayerEnters() (gas: 69051)
[PASS] testFulFillRandamWordsCanOnlyCalledAfterPerformUpKeep(uint256) (runs: 256, μ: 80837, ~: 80837)
[PASS] testFulfillRandomWordsAndPickWinnerAndSendMoney() (gas: 369870)
[PASS] testLotteryInLiveDuration() (gas: 13050)
[PASS] testLotteryInitializesInOpenState() (gas: 13178)
[PASS] testLotteryInitializesWithCorrectCallbackGasLimit() (gas: 13127)
[PASS] testLotteryInitializesWithCorrectEntranceFee() (gas: 13017)
[PASS] testLotteryInitializesWithCorrectGasLane() (gas: 13138)
[PASS] testLotteryInitializesWithCorrectRequestConfirmations() (gas: 13054)
[PASS] testLotteryRevertWhenYouDontPayEnoughEntranceFee() (gas: 22350)
[PASS] testLotteryRevertWhenYouDontPayEntranceFee() (gas: 13481)
[PASS] testLotterySavesPlayerWhenTheyEnter() (gas: 68623)
[PASS] testPerformUpKeepEmitsRequestId() (gas: 217341)
[PASS] testPerformUpKeepIfCheckUpKeepReturnsTrueUpdatesState() (gas: 214780)
[PASS] testRevertPerformUpKeepIfCheckUpKeepReturnsFalse() (gas: 72591)
Suite result: ok. 21 passed; 0 failed; 0 skipped; finished in 24.73ms (35.48ms CPU time)

Ran 1 test suite in 27.21ms (24.73ms CPU time): 21 tests passed, 0 failed, 0 skipped (21 total tests)
```

output of error 

```
forge script script/DeployLottery.s.sol:DeployLottery --rpc-url http://localhost:8545 --account Account1 --broadcast
[⠒] Compiling...
No files changed, compilation skipped
0x4e487b710000000000000000000000000000000000000000000000000000000000000011
EIP-3855 is not supported in one or more of the RPCs used.
Unsupported Chain IDs: 31337.
Contracts deployed with a Solidity version equal or higher than 0.8.20 might not work properly.
For more information, please see https://eips.ethereum.org/EIPS/eip-3855
Traces:
  [21607918] → new DeployLottery@0x5b73C5498c1E3b4dbA84de0F1833c4a029d90519
    └─ ← [Return] 107717 bytes of code

  [13285645] DeployLottery::run()
    ├─ [4501993] → new HelperConfig@0xC7f2Cf4845C6db0e1a1e91ED41Bcd0FcC1b0E141
    │   └─ ← [Return] 21407 bytes of code
    ├─ [3822626] HelperConfig::getConfig()
    │   ├─ [746241] → new LinkToken@0x7AaCc5300ec7Ac58fe86645D08f21b1BEcadf99a
    │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: HelperConfig: [0xC7f2Cf4845C6db0e1a1e91ED41Bcd0FcC1b0E141], amount: 1000000000000000000000000 [1e24])
    │   │   └─ ← [Return] 3267 bytes of code
    │   ├─ [0] VM::startBroadcast(DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38])
    │   │   └─ ← [Return] 
    │   ├─ [2863541] → new VRFCoordinatorV2_5Mock@0x34A1D3fff3958843C43aD80F30b94c510645C316
    │   │   ├─ emit ConfigSet()
    │   │   └─ ← [Return] 13951 bytes of code
    │   ├─ [0] VM::stopBroadcast()
    │   │   └─ ← [Return] 
    │   └─ ← [Return] NetworkConfig({ entranceFee: 10000000000000000 [1e16], lotteryDuration: 30, vrfCoordinator: 0x34A1D3fff3958843C43aD80F30b94c510645C316, gasLane: 0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c, subscriptionId: 0, requestConfirmations: 4, callbackGasLimit: 500000 [5e5], linkTokenAddress: 0x7AaCc5300ec7Ac58fe86645D08f21b1BEcadf99a, account: 0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38 })
    ├─ [4880454] → new CreateSubscription@0xdaE97900D4B184c5D2012dcdB658c008966466DD
    │   └─ ← [Return] 24263 bytes of code
    ├─ [6957] CreateSubscription::createSubscription(VRFCoordinatorV2_5Mock: [0x34A1D3fff3958843C43aD80F30b94c510645C316], DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38])
    │   ├─ [0] console::log("Creating a subscription on ChainId:", 31337 [3.133e4], DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38]) [staticcall]
    │   │   └─ ← [Stop] 
    │   ├─ [0] VM::startBroadcast(DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38])
    │   │   └─ ← [Return] 
    │   ├─ [2616] VRFCoordinatorV2_5Mock::createSubscription()
    │   │   └─ ← [Revert] panic: arithmetic underflow or overflow (0x11)
    │   └─ ← [Revert] panic: arithmetic underflow or overflow (0x11)
    └─ ← [Revert] panic: arithmetic underflow or overflow (0x11)



== Logs ==
  Creating a subscription on ChainId: 31337 0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38
Error: 
script failed: panic: arithmetic underflow or overflow (0x11)
make: *** [Makefile:34: deploy-anvil] Error 1
```

**Reply by EngrPips** [reply]
I meant the deployment script lol, sorry about that.

**Reply by EngrPips** [reply]
@saikrishna5907, In the output, when it creates a subscription for me, I have the below 

```js
 ├─ [52618] CreateSubscription::createSubscription(VRFCoordinatorV2PlusMock: [0x0165878A594ca255338adfa4d48449f69242Eb8F], 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76])
```

But for you, it shows the below. 

```js
├─ [6957] CreateSubscription::createSubscription(VRFCoordinatorV2_5Mock: [0x34A1D3fff3958843C43aD80F30b94c510645C316], DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38])
```

The bottom line is for your deployment process: when you called createSubscription, you called the function with a `VRFCoordinatorV2_5Mock` address and an address instead of passing in a privateKey. Can you figure out why is that?

**Reply by saikrishna5907** [reply]
Hi @EngrPips  sorry for late reply

Yes, recent course 2024 update @PatrickAlphaC suggested to use address instead, I think using private key is not secure.

**Reply by EngrPips** [reply]
Oh, I see. I haven't tried using address. I will try that sometime and let you know how it goes for me. Please show the `createSubscription` function inside of your `VRFCoordinatorV2_5Mock`. Let's see what it implements.

**Reply by EngrPips** [reply]
Hello @saikrishna5907, Did you resolve this already?

**Reply by saikrishna5907** [reply]
Hi @EngrPips  sorry I have been caught up with my work stuff, not yet I am using hardhat version project for now while I develop UI

**Reply by EngrPips** [reply]
Alright buddy, anytime you are ready we can try look into this issue again.

**Reply by nayashi002** [reply]
Am getting same error any solutions yet

**Reply by EngrPips** [reply]
Hello @nayashi002, I will advise that you open the `VRFCoordinatorV2_5Mock` contract and use `console.log` to find the point in the `CreateSubscription` function of the `VRFCoordinatorV2_5Mock` where the `overflow/underflow` is happening.

**Reply by lautaro3870** [reply]
Hello. I am having the same problem deploying the contract on local using foundry. I have the function as @EngrPips shows up but it is not working. Any idea i can try? Thankss

**Reply by EngrPips** [reply]
Hello @saikrishna5907 , Please check this discussion https://github.com/Cyfrin/foundry-full-course-cu/discussions/2246 I believe it will help you
